typing/ctype.ml

```diff
@@ -1837,8 +1834,8 @@ let enter_poly env univar_pairs t1 tl1 t2 tl2 f =
   let cl1 = List.map (fun t -> t, ref None) tl1
   and cl2 = List.map (fun t -> t, ref None) tl2 in
   univar_pairs := (cl1,cl2) :: (cl2,cl1) :: old_univars;
-  try let res = f t1 t2 in univar_pairs := old_univars; res
-  with exn -> univar_pairs := old_univars; raise exn
+  Misc.try_finally (fun () -> f t1 t2)
+    ~always:(fun () -> univar_pairs := old_univars)
```

Comment:
`Misc.protect_refs`.

Reply:
Not sure about this one. A priori, the code between the definition of `old_univars` and the call to `f` may be mutating `univar_pairs`, in which the case using `Misc.protect_refs` would not be correct.

Reply:
That's a fair point. Two remarks:

1. We could use `univar_paris := old_univars` and *then* call `Misc.protect_refs`, but that would be evil.
2. I don't know the type-checker better than I do, so I can't say whether the code above can reach this point with `!univar_paris <> old_univars`, but I can say with confidence that this code would be extremely fishy if it could, and probably wrong. So maybe it's worth asking someone that does do type-checker stuff to confirm here. (@trefis?)